packages/adp-tooling/src/preview/routes-handler.ts

```diff
@@ -0,0 +1,200 @@
+     * @param res Response
+     * @param next Next Function
+     */
+    public handleGetXMLFragmentByName = async (req: Request, res: Response, next: NextFunction) => {
+        try {
+            const xmlName = req.params['xmlName'];
+
+            if (!xmlName) {
+                res.send(HttpStatusCodes.NOT_FOUND).send(`Could not find XML Fragment at ${req.path}`);
+                return;
+            }
+
+            const xmlFragmentPath = path.join(__dirname, '../../templates/rta/ui', xmlName);
+
+            if (!fs.existsSync(xmlFragmentPath)) {
+                res.send(HttpStatusCodes.NOT_FOUND).send(`XML Fragment does not exist`);
+                return;
+            }
+
+            const fragment = fs.readFileSync(xmlFragmentPath, 'utf-8');
+
+            res.status(HttpStatusCodes.OK).contentType('application/xml').send(fragment);
+        } catch (e) {
+            res.status(HttpStatusCodes.INTERNAL_SERVER_ERROR).send(e.message);
+            next(e);
+        }
+    };
+
+    /**
+     * @description Handler for reading all fragment files from the workspace
+     * @param _ Request
+     * @param res Response
+     * @param next Next Function
+     */
+    public handleReadAllFragments = async (_: Request, res: Response, next: NextFunction) => {
+        const files = await this.project.byGlob('/**/changes/**/*.fragment.xml');
+        const xmlChangeFiles = await this.project.byGlob('/**/changes/*_addXML.change');
+
+        if (!files || files.length === 0) {
+            res.status(HttpStatusCodes.OK)
+                .contentType('application/json')
+                .send({ fragments: [], message: `No fragments found in the project workspace.` });
+        }
+        try {
+            // Dictionary where keys are XML Fragment names that are in the project's workspace
+            const dict: { [key: string]: string } = {};
+
+            const fragments = files.map((file) => {
+                const fileName = file.getName();
+                dict[fileName] = file.getPath();
+                return {
+                    fragmentName: fileName
+                };
+            });
+
+            /**
+             * Searches every addXML change file and deletes the fragmentName from the dictionary
+             */
+            await Promise.all(
+                xmlChangeFiles
+                    .map(async (file) => {
+                        const buffer = await file.getBuffer();
+                        const data: object & { content: { fragmentPath: string } } = JSON.parse(buffer.toString());
+                        // check if the fragment file exists in the workspace, return nothing if it does not
+                        const fragmentFileName = data.content.fragmentPath;
+                        const fragmentName = fragmentFileName.split('/').pop() as string;
+
+                        delete dict[fragmentName];
+                    })
+                    .map((f) => Promise.resolve(f))
```

Comment:
Because otherwise the `dict` object will be full and no keys will be deleted. That's why you need to wait for the promise to resolve. Also, I will be deleting this functionality since the new dialog doesn't require this.

Reply:
Ok, if you delete this one, I won't argue further and mark this conversation as resolved the moment it is gone.